File: src/main/java/org/meetkt/basket/domain/model/Items.java
package org.meetkt.basket.domain.model;

import org.meetkt.catalogue.domain.model.ProductId;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Items extends ArrayList<Item> {

    public Items(List<Item> items) {
        super(items);
    }

    public Optional<Item> findFirst(ProductId productId) {
        return this
                .stream()
                .filter(item -> item.contains(productId))
                .findFirst();
    }

    public Price totalPrice() {
        return this.
                stream()
                .map(Item::price)
                .reduce(Price.ZERO, Price::add);
    }
}

File: src/main/kotlin/in/meetkt/basket/domain/model/Items.kt
package `in`.meetkt.basket.domain.model

import `in`.meetkt.catalogue.domain.model.ProductId

class Items(private val items: MutableList<Item>) : MutableList<Item> by items {

    fun totalPrice(): Price =
        items.map { it.price }.foldRight(Price.ZERO, { accumulator, price -> accumulator.add(price) })

    fun findFirst(productId: ProductId): Item? = items.firstOrNull { item -> item.contains(productId) }

    override fun toString(): String {
        return "Items(items=$items)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Items
        if (items != other.items) return false
        return true
    }

    override fun hashCode(): Int {
        return items.hashCode()
    }
}